statsrelay/src/backend_client/prom_client.rs

```diff
@@ -0,0 +1,522 @@
+    ) -> Result<reqwest::Response, reqwest::Error> {
+        self.inner
+            .post(self.endpoint.clone())
+            .header("Content-Type", "application/x-protobuf")
+            .header("Content-Encoding", "snappy")
+            .header("X-Prometheus-Remote-Write-Version", "0.1.0")
+            .body(compressed_write_request)
+            .send()
+            .await
+    }
+}
+
+async fn client_task<T: PromWriteRequestClient + Send + Sync + 'static>(
+    stats: Stats,
+    client: Arc<T>,
+    mut rx: mpsc::Receiver<Pdu>,
+    trip: Tripwire,
+    config: PromBackendClientConfig,
+) {
+    let semaphore = Arc::new(Semaphore::new(config.max_in_flight));
+    // The receiving end of the flush channel blocks until all spawned tasks have terminated.
+    let (flush_tx, mut flush_rx) = mpsc::channel::<()>(1);
+
+    loop {
+        let (batch, shutdown) = construct_batch(
+            &mut rx,
+            Duration::from_millis(config.batch_creation_timeout_ms),
+            config.max_batch_size,
+        )
+        .await;
+
+        if !batch.is_empty() {
+            let permit = semaphore.clone().acquire_owned().await.unwrap();
+            stats
+                .requests_in_flight
+                .set((config.max_in_flight - semaphore.available_permits()) as f64);
+
+            let stats = stats.clone();
+            let retry_delay_ms = config.retry_delay_ms;
+            let client = client.clone();
+            let flush_tx = flush_tx.clone();
+            let trip = trip.clone();
+
+            tokio::spawn(async move {
+                let samples_in_batch = batch.len() as f64;
+                stats.requests_total.inc();
+                stats.samples_total.inc_by(samples_in_batch);
+
+                let write_request = construct_write_request(&stats, batch);
+                // N.B. According to the snap documentation, this compession should never fail for
+                // our use case.
+                let compressed_write_request = match compress_write_request(write_request) {
+                    Ok(c) => c,
+                    Err(e) => {
+                        stats.requests_compression_fail.inc();
+                        error!(
+                            "failed to compress prometheus WriteRequest, dropping samples: {:?}",
+                            e
+                        );
+                        return;
+                    }
+                };
+                stats
+                    .bytes_sent_total
+                    .inc_by(compressed_write_request.len() as f64);
+
+                // Attempt each request at least once
+                let mut res = client
+                    .send_write_request(compressed_write_request.clone())
+                    .await;
+                // And retry only if the tripwire has not yet been triggered.
+                while should_retry(&res) {
+                    sleep(Duration::from_millis(retry_delay_ms)).await;
```

Comment:
Don't have to do it right now, but some jitter and/or backoff might be a nice future addition to the retry mechanism. You might also be able to incorporate a more general retry mechanism at a higher layer in this stack -- say by wrapping at the `BackendClient` layer with a `RetryingBackendClient : BackendClient` or similar -- to avoid having to implement it separately in each backend that's able to fail.